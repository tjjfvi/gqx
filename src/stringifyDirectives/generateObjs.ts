
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
