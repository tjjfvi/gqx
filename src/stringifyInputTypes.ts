
import stringifyInputType from "./stringifyInputType";
import { Context } from ".";

export default (ctx: Context) =>
  Object.entries(ctx.inputTypes).map(([name, fields]) => {
    ctx.exports.push(name);
    return stringifyInputType(name, fields)
  }).join("\n\n")
