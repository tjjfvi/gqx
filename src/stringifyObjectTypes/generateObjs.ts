
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
