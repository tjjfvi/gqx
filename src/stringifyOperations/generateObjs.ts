
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
