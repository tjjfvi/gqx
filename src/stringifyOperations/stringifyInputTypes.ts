
import { Obj } from ".";
import stringifyInputType from "../stringifyInputType";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[]) =>
  stringifyNamespace("$$Input", objs, (o, p) => stringifyInputType(p.id.prop, p.args))
