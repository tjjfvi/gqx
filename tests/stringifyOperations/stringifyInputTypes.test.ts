import generateObjs from "../../src/stringifyOperations/generateObjs"
import stringifyInputTypes from "../../src/stringifyOperations/stringifyInputTypes";

test("", () => {
  let [objs] = generateObjs(utils.getOperations(["stringifyOperations"]));
  let code = stringifyInputTypes(objs);
  expect(code).toParse().toMatchSnapshot();
})

