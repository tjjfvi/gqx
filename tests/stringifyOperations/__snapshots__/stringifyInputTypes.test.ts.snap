// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export namespace $$Input {
  export namespace Query {
    export interface argsObjectArr2D {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrMM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrMN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrNM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrNN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectMand {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectNull {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArr2D {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArrMM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArrMN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArrNM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArrNN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarMand {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarNull {
      a: String;
      b?: (Float | null | void);
    }
    export interface noArgsObjectArr2D {}
    export interface noArgsObjectArrMM {}
    export interface noArgsObjectArrMN {}
    export interface noArgsObjectArrNM {}
    export interface noArgsObjectArrNN {}
    export interface noArgsObjectMand {}
    export interface noArgsObjectNull {}
    export interface noArgsScalarArr2D {}
    export interface noArgsScalarArrMM {}
    export interface noArgsScalarArrMN {}
    export interface noArgsScalarArrNM {}
    export interface noArgsScalarArrNN {}
    export interface noArgsScalarMand {}
    export interface noArgsScalarNull {}
  }
  export namespace Mutation {
    export interface argsObjectArr2D {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrMM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrMN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrNM {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectArrNN {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectMand {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsObjectNull {
      a: String;
      b?: (Float | null | void);
    }
    export interface argsScalarArr2D {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarArrMM {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarArrMN {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarArrNM {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarArrNN {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarMand {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface argsScalarNull {
      a: Boolean;
      b?: (Int | null | void);
    }
    export interface noArgsObjectArr2D {}
    export interface noArgsObjectArrMM {}
    export interface noArgsObjectArrMN {}
    export interface noArgsObjectArrNM {}
    export interface noArgsObjectArrNN {}
    export interface noArgsObjectMand {}
    export interface noArgsObjectNull {}
    export interface noArgsScalarArr2D {}
    export interface noArgsScalarArrMM {}
    export interface noArgsScalarArrMN {}
    export interface noArgsScalarArrNM {}
    export interface noArgsScalarArrNN {}
    export interface noArgsScalarMand {}
    export interface noArgsScalarNull {}
  }

  export type $<T> =
    | (T extends typeof Query$argsObjectArr2D ? Query.argsObjectArr2D : never)
    | (T extends typeof Query$argsObjectArrMM ? Query.argsObjectArrMM : never)
    | (T extends typeof Query$argsObjectArrMN ? Query.argsObjectArrMN : never)
    | (T extends typeof Query$argsObjectArrNM ? Query.argsObjectArrNM : never)
    | (T extends typeof Query$argsObjectArrNN ? Query.argsObjectArrNN : never)
    | (T extends typeof Query$argsObjectMand ? Query.argsObjectMand : never)
    | (T extends typeof Query$argsObjectNull ? Query.argsObjectNull : never)
    | (T extends typeof Query$argsScalarArr2D ? Query.argsScalarArr2D : never)
    | (T extends typeof Query$argsScalarArrMM ? Query.argsScalarArrMM : never)
    | (T extends typeof Query$argsScalarArrMN ? Query.argsScalarArrMN : never)
    | (T extends typeof Query$argsScalarArrNM ? Query.argsScalarArrNM : never)
    | (T extends typeof Query$argsScalarArrNN ? Query.argsScalarArrNN : never)
    | (T extends typeof Query$argsScalarMand ? Query.argsScalarMand : never)
    | (T extends typeof Query$argsScalarNull ? Query.argsScalarNull : never)
    | (T extends typeof Query$noArgsObjectArr2D ? Query.noArgsObjectArr2D : never)
    | (T extends typeof Query$noArgsObjectArrMM ? Query.noArgsObjectArrMM : never)
    | (T extends typeof Query$noArgsObjectArrMN ? Query.noArgsObjectArrMN : never)
    | (T extends typeof Query$noArgsObjectArrNM ? Query.noArgsObjectArrNM : never)
    | (T extends typeof Query$noArgsObjectArrNN ? Query.noArgsObjectArrNN : never)
    | (T extends typeof Query$noArgsObjectMand ? Query.noArgsObjectMand : never)
    | (T extends typeof Query$noArgsObjectNull ? Query.noArgsObjectNull : never)
    | (T extends typeof Query$noArgsScalarArr2D ? Query.noArgsScalarArr2D : never)
    | (T extends typeof Query$noArgsScalarArrMM ? Query.noArgsScalarArrMM : never)
    | (T extends typeof Query$noArgsScalarArrMN ? Query.noArgsScalarArrMN : never)
    | (T extends typeof Query$noArgsScalarArrNM ? Query.noArgsScalarArrNM : never)
    | (T extends typeof Query$noArgsScalarArrNN ? Query.noArgsScalarArrNN : never)
    | (T extends typeof Query$noArgsScalarMand ? Query.noArgsScalarMand : never)
    | (T extends typeof Query$noArgsScalarNull ? Query.noArgsScalarNull : never)
    | (T extends typeof Mutation$argsObjectArr2D ? Mutation.argsObjectArr2D : never)
    | (T extends typeof Mutation$argsObjectArrMM ? Mutation.argsObjectArrMM : never)
    | (T extends typeof Mutation$argsObjectArrMN ? Mutation.argsObjectArrMN : never)
    | (T extends typeof Mutation$argsObjectArrNM ? Mutation.argsObjectArrNM : never)
    | (T extends typeof Mutation$argsObjectArrNN ? Mutation.argsObjectArrNN : never)
    | (T extends typeof Mutation$argsObjectMand ? Mutation.argsObjectMand : never)
    | (T extends typeof Mutation$argsObjectNull ? Mutation.argsObjectNull : never)
    | (T extends typeof Mutation$argsScalarArr2D ? Mutation.argsScalarArr2D : never)
    | (T extends typeof Mutation$argsScalarArrMM ? Mutation.argsScalarArrMM : never)
    | (T extends typeof Mutation$argsScalarArrMN ? Mutation.argsScalarArrMN : never)
    | (T extends typeof Mutation$argsScalarArrNM ? Mutation.argsScalarArrNM : never)
    | (T extends typeof Mutation$argsScalarArrNN ? Mutation.argsScalarArrNN : never)
    | (T extends typeof Mutation$argsScalarMand ? Mutation.argsScalarMand : never)
    | (T extends typeof Mutation$argsScalarNull ? Mutation.argsScalarNull : never)
    | (T extends typeof Mutation$noArgsObjectArr2D ? Mutation.noArgsObjectArr2D : never)
    | (T extends typeof Mutation$noArgsObjectArrMM ? Mutation.noArgsObjectArrMM : never)
    | (T extends typeof Mutation$noArgsObjectArrMN ? Mutation.noArgsObjectArrMN : never)
    | (T extends typeof Mutation$noArgsObjectArrNM ? Mutation.noArgsObjectArrNM : never)
    | (T extends typeof Mutation$noArgsObjectArrNN ? Mutation.noArgsObjectArrNN : never)
    | (T extends typeof Mutation$noArgsObjectMand ? Mutation.noArgsObjectMand : never)
    | (T extends typeof Mutation$noArgsObjectNull ? Mutation.noArgsObjectNull : never)
    | (T extends typeof Mutation$noArgsScalarArr2D ? Mutation.noArgsScalarArr2D : never)
    | (T extends typeof Mutation$noArgsScalarArrMM ? Mutation.noArgsScalarArrMM : never)
    | (T extends typeof Mutation$noArgsScalarArrMN ? Mutation.noArgsScalarArrMN : never)
    | (T extends typeof Mutation$noArgsScalarArrNM ? Mutation.noArgsScalarArrNM : never)
    | (T extends typeof Mutation$noArgsScalarArrNN ? Mutation.noArgsScalarArrNN : never)
    | (T extends typeof Mutation$noArgsScalarMand ? Mutation.noArgsScalarMand : never)
    | (T extends typeof Mutation$noArgsScalarNull ? Mutation.noArgsScalarNull : never)
}"
`;
